### ISSUE TITLE
Four sm_12x items in the GLM-5.3 vLLM stack (not in cuda-exl3), all present unpatched in our image — from tpurtell's patch set

### ISSUE THREAD

**Issue by Zeuss5** [Zeuss5/cuda-exl3#6]
Not a cuda-exl3 bug report — this is stack-level, outside the plugin, and it lands on **both** our parts because the gate that matters keys on `major >= 9` and GB10 is sm_121, `major = 12`, exactly as sm_120 is.

Source: `tpurtell/glm-5.3-flash-ext3-2x-rtx` (Apache-2.0), which the author pointed me at. Its EXL3 kernels are B12x's and closed — the patch set is against a private vLLM fork — but `patches/port-glm53-sm12-stability.py` and `patches/port-glm-prefill-jit.py` are findings about **near-stock vLLM code**, and I checked each one against the image we actually serve from (`vllm 0.1.dev20111+g7f1e92bec`). Credit for all four is theirs; the verification below is mine.

---

## 1. Programmatic Dependent Launch is enabled on sm_12x, and they found KDA state races there

`vllm/platforms/cuda.py`:

```python
    @classmethod
    def is_arch_support_pdl(cls) -> bool:
        ...
        return major >= 9
```

Their change is `return major in (9, 10)`, with the note: *"The day-zero gate enables Programmatic Dependent Launch on every capability >= 9. The closely related dual-Spark qualification found KDA recurrent-state races on SM12x, where PDL has not been validated."*

Confirmed live on our card:

```
capability: (12, 0)
is_arch_support_pdl(): True
MHC ENABLE_PDL: True
```

The call sites this turns on are **all hot in GLM-5.3-Flash**, not incidental:

* `model_executor/layers/mamba/ops/causal_conv1d.py` (×2) — the KDA conv
* `model_executor/layers/mamba/ops/gather_initial_states.py` — **the recurrent-state gather, which is the thing they say races**
* `third_party/flash_linear_attention/ops/fused_norm_gate.py` (×2) — KDA gating
* `model_executor/kernels/mhc/tilelang_kernels.py` — `ENABLE_PDL`, i.e. the hyper-connection mixing that is **10.6 % of your prefill**

34 of 45 GLM-5.3-Flash layers are KDA and `hc_mult = 4`, so every one of these fires on every step. **I am A/B-ing PDL on vs off end to end here right now** (throughput plus a 12-run greedy determinism check) and will post numbers. If it is free, it is cheap insurance on a path where someone else has already seen corruption; if it costs throughput, that is a real trade to describe rather than a one-line fix.

## 2. Uninitialised `pool_topk` reaches the cache as pool ids

`model_executor/layers/sparse_attn_indexer_kpool.py`, **three** sites in our image (lines 939, 1019, 1137):

```python
pool_topk = torch.empty((score_rows, work_k), dtype=torch.int32, device=...)
```

Their note: *"The top-k kernels only promise `min(k, valid)` outputs. Initializing the remainder makes short/local DCP rows deterministically invalid instead of allowing uninitialized pool ids to address unwritten cache rows."* Fix is `torch.full(..., -1)`.

This is the same failure shape as the EP bug you found in #1 — a kernel that writes a prefix, and a consumer that reads the whole array. Their anchors say `select_k` where ours say `work_k`, so the code has drifted, but the defect is at all three sites. They attribute it to short/local **DCP** rows; you are TP=3 + EP, so this may not fire for you, but it is uninitialised memory addressing a cache either way.

## 3. Missing upper bound in the kpool compress kernel

`models/glm5next/nvidia/ops/kpool_compress.py:1053`:

```python
hist_out = tl.where(pid >= 0, hist_val, -1)          # ours
hist_out = tl.where((pid >= 0) & (pid < pool_len), hist_val, -1)   # theirs
```

Lower-bounded only. Same family as 2.

## 4. The DSA indexer's prefill-metadata kernel specialises on chunk offsets

`v1/attention/backends/mla/indexer.py` takes `query_slice_start: int` / `query_slice_stop: int` as Triton kernel scalars (lines 281–282, 370–371) with **no** `do_not_specialize`. Triton specialises int args on value and on 16-alignment, so distinct chunk offsets recompile the kernel.

What makes this look like an oversight rather than a choice: the sibling backends in the *same tree* already do exactly this — `v1/attention/backends/b12x_attn.py` uses `@triton.jit(do_not_specialize=["num_reqs"])` twice, and `mla/sparse_swa.py` uses it for `num_prefills`, `num_decodes`, `window_size`, `token_offset`. Only the indexer omits it.

Their fix adds both `do_not_specialize` and `do_not_specialize_on_alignment` for the two offsets. You measured the indexer at **0.5 % of prefill** so its *kernel* time is closed — but this is compile time, not kernel time, and it would show up as tail-pass jitter and long-prefill stalls rather than in the steady-chunk breakdown. It may be worth a look next to the MLA-autotuner re-tune you already identified as inflating your window average.

---

## One more, which I could not confirm here and you should ignore unless you run DCP

They also fix a Mamba block table that is sharded by `dcp_size` although Mamba state is replicated (46 columns at 300 K where the GDN builder needed 91, crashing at column 47), and an FLA chunk-segmentation staleness where `prepare_chunk_indices` is memoised by `tensor_cache`, which matches on **`a is b`** — object identity, not contents — so a reused `cu_seqlens` buffer returns the previous chunk's segmentation. vLLM's own source carries the comment *"This will be fixed by https://github.com/vllm-project/vllm/pull/51540"*, and their patch is a port of that PR into GLM's KDA wrapper.

Our image has neither the DCP anchor nor `non_spec_chunk_indices`, and I have not established that the identity collision actually fires in a TP-only configuration — a fresh `buf[:n]` slice is a new Python object each build, so the cache should miss rather than go stale. Recording it because it is a **silent wrong-answer** class if it does fire, and you change chunk size under KV pressure more than I do.

**Comment by NNNtrance** [comment]
Checked all four against both images we serve from (the EXL3 image on vLLM `0.1.dev20051+g487ecf187`, and the NVFP4 image on the lab fork `0.1.dev0+lil.jovian.9c4dd0548`), patches written and dry-tested, A/B queued behind the measurements already on the GPU. Two corrections to the issue text from that work, both small:

**Item 4 is one cold JIT per engine process, not a recompile per chunk offset.** Against the image's own triton 3.7.1, an integer scalar argument has exactly three specialisation classes: `== 1`, `% 16 == 0`, and "other". The engine's warm-up already covers all three for `query_slice_stop` and misses only the "other" class for `query_slice_start`, which `_build_chunk_specs` reaches once (`max_q = max_logits_elems // chunk_n`). So the symptom is a single long-prefill stall after start-up, not steady-state throughput or per-chunk jitter. `do_not_specialize` still removes it; it just buys less than the issue implies.

**Item 1 has one more call site on the EXL3 image:** `model_executor/layers/mamba/ops/scatter_states.py` also launches under the PDL flag (absent in the NVFP4 fork). Status of the rest: PDL gate present unpatched in both (`cuda.py:758` / `:718`, byte-identical); item 2 present at two sites in ours (`select_k` anchors, 538 and 803 — you count three, tpurtell two; the code has drifted) and absent in the NVFP4 fork (buffer moved into the model file); item 3 present at line 938 in ours, absent in the fork, where the reader is a closed kernel; item 5 inert at DCP=1 in ours and already fixed in the fork (`dcp_replicated` guard); item 6 does not fire in either (fresh slices each build).

Plan: baseline determinism first (6 prompts × 12 greedy runs, byte compare) on the untouched production tree, then one dump boot with all four applied, then the combined arm, and PDL alone only if the combined arm moves anything. On our stack adding a patch file changes the fast-load sidecar identity, so the arms are env-selected after one re-dump rather than one boot per patch. Numbers will follow here.

**Comment by Zeuss5** [comment]
Both corrections accepted, and thank you for checking the specialisation classes rather than taking my word for it — **item 4 is one cold JIT per engine process, not per-chunk recompiles**, and I should have enumerated triton's three integer classes before writing "distinct chunk offsets recompile the kernel". That was inference from the missing decorator plus the sibling backends using it, not a measurement. `scatter_states.py` noted as a fifth PDL call site.

## Before you run the determinism baseline — the plugin itself is not bit-exact, and it is my fault, not PDL's

I ran that experiment here first and it came back **24/24 distinct completions out of 24 greedy runs, temperature 0, one request at a time — in _both_ arms, PDL on and PDL off.** So the test as I specified it has no discriminating power at all, and if you run your 6 × 12 byte-compare on the untouched tree you will get the same and it will mean nothing about PDL.

The cause is in cuda-exl3, and it is a bug I have just fixed:

* `README.md` has claimed since the beginning that `CUDA_EXL3_DETERMINISTIC=1` is "bit-exact everywhere".
* It only ever disabled split-k on the **dense** path (`ops.py:113`).
* `5814c7f` folded the combine into the down projection's epilogue, and that epilogue accumulates each routed row into its token's row with `atomicAdd` (`exl3_had.cuh:300`). Nothing gates it.

So on **every MoE model** — which is every model either of us serves — the flag has never delivered what it promised, and the MoE stage has been non-bit-exact run to run whether or not it was set. Fix routes the deterministic path back through `exl3_moe_combine`, which is still built, still bound and still covered by `test_exl3_moe_split.py`; it sums each token's top-k in a fixed `k` order. Committing shortly.

### What this means for your plan

`CUDA_EXL3_DETERMINISTIC=1` alone still will not give you a byte-stable baseline, because **your TP all-reduce is not bit-exact either** and no flag of mine changes that. Two ways through, in order of what I would try:

1. **Drop determinism as the PDL probe entirely and use divergence instead.** Fix the seed and the prompt, run the model twice per arm, and compare *logits* at a fixed token position rather than sampled text — a race shows up as a magnitude far outside the fp32-atomics-plus-all-reduce noise floor you can characterise from the two runs *within* one arm. That noise floor is the measurement; text identity was never going to be.
2. **If you want a genuinely byte-stable baseline**, you need `CUDA_EXL3_DETERMINISTIC=1` *and* TP=1 *and* the fused epilogue off — which is a configuration nobody serves, so it tests the kernels rather than the stack.

My throughput half is worth having regardless, since it is the half that decides whether the fix is free. 4× RTX PRO 6000, TP=4, same image, PDL on vs off:

| | PDL on | PDL off | Δ |
|---|---|---|---|
| prefill TTFT (4 k) | 2 553 ms | 2 515 ms | −1.5 % (off faster) |
| C1 output tok/s | 93.29 | 93.07 | −0.2 % |
| C1 TPOT | 9.12 ms | 9.14 ms | +0.2 % |
| C4 output tok/s | 136.02 | 133.87 | −1.6 % |

Everything is inside run-to-run noise, and prefill — where the MHC TileLang kernels that take the flag are 10.6 % of your step — is nominally *faster* with PDL off. **On sm_120 PDL buys nothing measurable**, so if tpurtell's race is real the fix costs nothing to take. Whether it is real is what your logit-divergence arm can still answer; my determinism arm cannot, and I would rather say so before you spend a boot on it.

**Comment by Zeuss5** [comment]
Fix is in — `e7e345e`, pushed. Under `CUDA_EXL3_DETERMINISTIC=1` the down projection now writes routed rows and `exl3_moe_combine` sums each token's top-k in a fixed `k` order. Three tests added; suite is 101 green.

## The number you need for the logit-divergence arm

Chasing the fix turned up something that bears directly on your noise floor. The fused epilogue accumulates through **bf16** atomics, so it rounds once per routed row; the combine sums in fp32 and rounds once per token. Against an fp64 reference:

| top_k | fused epilogue | `exl3_moe_combine` |
|---|---|---|
| 1 | 2.38e-3 | 1.66e-3 |
| 4 | 3.53e-3 | 1.66e-3 |
| **8** (GLM-5.3-Flash) | **4.34e-3** | 1.67e-3 |
| 16 | 5.51e-3 | 1.66e-3 |

The combine is flat because 1.7e-3 is just the bf16 rounding of the stored result. The fused path grows as the partial sums round.

**So your MoE stage has been contributing ~4.3e-3 of relative error per layer, and a run-to-run varying part of it, on top of the all-reduce.** If you were going to look for a PDL race as a logit deviation, that is the floor you would have been measuring against. Running the divergence arm with `CUDA_EXL3_DETERMINISTIC=1` drops it to 1.7e-3 *and* removes our run-to-run component entirely, which should make a real race far easier to separate from arithmetic. It costs throughput (it reinstates a kernel and an `(rows, H)` tensor per MoE layer) — but for a diagnostic arm that is the right trade, and it makes the two runs within an arm differ only by the all-reduce.

I am **not** proposing that as a serving default. Accumulating the fused path in fp32 would reinstate exactly the buffer and the extra pass the fusion exists to remove, and MMLU does not resolve the difference here (86.32 ± 0.75 against 86.4 ± 0.7 with the fused path live), so it stays a trade rather than a bug. If your accuracy work ever does resolve it, that changes and I would want to know.

One thing I would still like from your side, unrelated to PDL and still the top in-plugin item from #5: **arm C of the expert-reread bench on a 48-SM part.** `9b17ea9` closed it here — the trellis is resident across an expert's blocks, and the per-block traffic model needs 145–151 % of the measured ruler, which is impossible — but the mechanism is that `moe_align_block_size` makes an expert's blocks consecutive in `blockIdx.y` so they co-schedule, and the concurrent-block window is four times narrower on 48 SMs than on 188. That is the one place where my "closed" genuinely might not transfer.

**Comment by NNNtrance** [comment]
Taking the divergence design as specified, and one of your two asks is already answered.

**Arm C on a 48-SM part — measured, and your closure transfers.** We ran `bench_moe_expert_reread.py` arm C on the GB10 with the production kernels (three runs, throwaway container, ruler in the same binary: 218–235 GB/s). Doubling the blocks per expert costs ×1.11 in time, not ×2; tripling ×1.20; quadrupling ×1.5 — and the per-block "bandwidth" reads 165–299 % of the ruler, which is only possible if the trellis stays resident across an expert's consecutive blocks. So on 48 SMs the co-scheduling window is narrower, as you said, but still wide enough: the expert's blocks re-use the trellis from L2, and the traffic model that needed 145–151 % of the ruler is impossible here too. Closed on both parts.

| N (blocks per expert) | time vs N=1 | per-block GB/s vs ruler |
|---|---|---|
| 2 | ×1.11 | 165–201 % |
| 3 | ×1.20 | 228–281 % |
| 4 | ×1.50 | 245–299 % |

**Determinism → divergence.** Agreed that byte identity was never going to work on a TP stack, and thank you for finding the epilogue atomics before we spent the boot. Revised arm: fixed prompt set and seed, two runs per arm, compare the per-position logprob vectors of the greedy continuation (all positions, not one), characterise the within-arm floor first, then look for outliers with PDL on vs off. The diagnostic arm runs on an image built at `e7e345e` with `CUDA_EXL3_DETERMINISTIC=1` so the within-arm difference is the all-reduce alone; production stays at `754421f` — the 4.3e-3 vs 1.7e-3 per-layer error is a trade we will put in front of our accuracy tools, and if anything resolves it you will hear.

Throughput half on the GB10 (PDL on vs off, TP=3 + EP, same boot) comes with the same arm. Your sm_120 numbers say the fix is free there; we will report whether 48 SMs agree.

**Comment by NNNtrance** [comment]
Measured on the three GB10 nodes, TP=3 + EP, the divergence design as agreed (fixed prompts and seed, two runs per arm, per-position logprobs of the greedy continuation, `CUDA_EXL3_DETERMINISTIC=1` on an image at `e7e345e` for the diagnostic arms, production `754421f` untouched as arm 0).

**No PDL race detected on sm_121.** Within-arm floor with the deterministic MoE stage: p95 |Δlogprob| 8.6e-2 (PDL on) / 7.8e-2 (PDL off), earliest token divergence at position 3 — that is the all-reduce alone, as you predicted. Between arms (PDL on vs off, first runs, K=4): max 2.9e-1 against a pooled within-arm p95 of 8.0e-2, between-arm p95 6.2e-2 — below the within-arm p95 — and the earliest between-arm divergence is at position 5, later than within an arm. No prompt flagged. For reference, production without the deterministic flag has a floor p95 of 1.07e-1 and diverges from itself at position 6.

**PDL on vs off is a wash on 48 SMs.** Two boots per setting; per-stream C1 ranges interleave (on: 69.4–79.7, off: 72.6–79.6), aggregate C1 within ~1 %, C8 within 1 %, prefill-fresh within 1 %. Same as your sm_120 result, so the gate change costs nothing here either.

**Items 2 and 3 (uninitialised pool ids, missing upper bound) and item 4 (indexer specialisation): free.** PDL off with the three items vs PDL off with them reverted: C1 70.9 vs 70.1 aggregate, C8 197.3 vs 199.2, prefill-fresh 1,748 vs 1,733 — all inside our noise band, gates 10/10 · 12/12 in every arm (one 11/12 warm reading on a known-flaky item repeated 12/12 three times). Item 4's one cold JIT is not measurable from the client: in every arm the first ~60K-token prefill run was the fastest or within 2 % of the rest; steady-state throughput unchanged, as expected.

We will carry all four as a correctness/insurance set into the next production configuration, since they cost nothing; the race itself we could not reproduce on this topology.
